File: logic/src/subs.cpp
#include "subs.h"

void Subs::follow(const std::string &follower, const std::string &followed) {
    auto f_it = all_subs.find(follower);

    //If follower has no subs create new list
    //If list exists, add to list
    if(f_it == all_subs.end()) {
        std::unordered_set<std::string> nset;
        nset.insert(followed);
        all_subs[follower] = nset;
    } else {
        f_it->second.insert(followed);
    }
}

std::unordered_set<std::string> Subs::followedBy(const std::string &follower) {
    std::unordered_set<std::string> ret;

    //Lookup users that the follower has subscribed
    auto f_it = all_subs.find(follower);
    if ( f_it != all_subs.end()) {
        ret = f_it->second;
    }

    //Add follower to list -- all users follow themselves
    ret.insert(follower);
    return ret;
}

File: java/codechallenge/src/main/java/codechallenge/impl/Subs.java
package codechallenge.impl;

import java.util.HashMap;
import java.util.HashSet;
import java.util.Map;
import java.util.Set;

import codechallenge.interfaces.SubsInterface;

public class Subs implements SubsInterface {
	private Map<String, Set<String>> all_subs;
	
	public Subs() {
		super();
		
		all_subs = new HashMap<>();
	}
	
	@Override
	public void follow(String follower, String followed) {
		//If follower has no subs create new list
	    //If list exists, add to list
		
		Set<String> usubs = all_subs.getOrDefault(follower, null);
		
		if(usubs == null) {
			usubs = new HashSet<>(1);
			usubs.add(followed);
			all_subs.put(follower, usubs);
		} else {
			usubs.add(followed);
		}
		
	}

	@Override
	public Set<String> followedBy(String follower) {
		//Lookup users that the follower has subscribed
		Set<String> usubs = all_subs.getOrDefault(follower, null);
		
		if(usubs == null) { usubs = new HashSet<>(1); }
		
		//Add follower to list -- all users follow themselves
		usubs.add(follower);
		
		return usubs;
	}

	
}
